internal/server/http/dump.go

```diff
@@ -0,0 +1,285 @@
+				buf = append(buf, lowerhex[s[0]&0xF])
+			case r > utf8.MaxRune:
+				r = 0xFFFD
+				fallthrough
+			case r < 0x10000:
+				buf = append(buf, `\u`...)
+				for s := 12; s >= 0; s -= 4 {
+					buf = append(buf, lowerhex[r>>uint(s)&0xF])
+				}
+			default:
+				buf = append(buf, `\U`...)
+				for s := 28; s >= 0; s -= 4 {
+					buf = append(buf, lowerhex[r>>uint(s)&0xF])
+				}
+			}
+		}
+	}
+	return buf
+}
+
+// Copied from the gorilla/handlers package
+func buildCommonLogLine(req *http.Request, url url.URL, ts time.Time, status int, size int) []byte {
+	username := "-"
+	if url.User != nil {
+		if name := url.User.Username(); name != "" {
+			username = name
+		}
+	}
+
+	host, _, err := net.SplitHostPort(req.RemoteAddr)
+	if err != nil {
+		host = req.RemoteAddr
+	}
+
+	uri := req.RequestURI
+
+	// Requests using the CONNECT method over HTTP/2.0 must use
+	// the authority field (aka r.Host) to identify the target.
+	// Refer: https://httpwg.github.io/specs/rfc7540.html#CONNECT
+	if req.ProtoMajor == 2 && req.Method == "CONNECT" {
+		uri = req.Host
+	}
+	if uri == "" {
+		uri = url.RequestURI()
+	}
+
+	buf := make([]byte, 0, 3*(len(host)+len(username)+len(req.Method)+len(uri)+len(req.Proto)+50)/2)
+	buf = append(buf, host...)
+	buf = append(buf, " - "...)
+	buf = append(buf, username...)
+	buf = append(buf, " ["...)
+	buf = append(buf, ts.Format("02/Jan/2006:15:04:05 -0700")...)
+	buf = append(buf, `] "`...)
+	buf = append(buf, req.Method...)
+	buf = append(buf, " "...)
+	buf = appendQuoted(buf, uri)
+	buf = append(buf, " "...)
+	buf = append(buf, req.Proto...)
+	buf = append(buf, `" `...)
+	buf = append(buf, strconv.Itoa(status)...)
+	buf = append(buf, " "...)
+	buf = append(buf, strconv.Itoa(size)...)
+	return buf
+}
+
+// copied from gorilla/handlers and modified to add the body
+// writeLog writes a log entry for req to w in Apache Common Log Format.
+// ts is the timestamp with which the entry should be logged.
+// status and size are used to provide the response HTTP status and size.
+func writeLog(writer io.Writer, params handlers.LogFormatterParams, body string) {
+	buf := buildCommonLogLine(params.Request, params.URL, params.TimeStamp, params.StatusCode, params.Size)
+	// Append body if present
+	if len(body) > 0 {
+		buf = append(buf, " "...)
+		buf = append(buf, body...)
+	}
+	buf = append(buf, '\n')
+	writer.Write(buf)
+}
+
+// isBinaryContent checks to see if the body is a common binary content type
+func isBinaryContent(r *http.Request) bool {
+	contentType := r.Header.Get("Content-Type")
+	for _, binaryType := range binaryContentTypes {
+		if strings.HasPrefix(contentType, binaryType) {
+			return true
+		}
+	}
+	return false
+}
+
+func shouldRecordRequest(s *Server) bool {
+	return len(s.dumpRequestsPath) > 0 && s.dumpCh != nil
+}
+
+func getBody(r *http.Request, s *Server) string {
+	if s.logLevel == 0 && !shouldRecordRequest(s) {
+		return ""
+	}
+	bodyBytes, err := io.ReadAll(r.Body)
```

Comment:
I'd love to see some limits here. Does it really make sense to write the body over and over, especially if large? 🤔  What's the purpose?

Reply:
We can definitely put a limit if needed. It would need to be configurable for when I want a very large body. What would you say a good limit would be?

Reply:
Added a io.LimitedReader to read up to a configurable maximum bytes now: https://github.com/friendsofgo/killgrave/pull/170/commits/78a8ae0658afdcc35cff382e006c1e2f315142ad